src/VisualStudio/Core/Def/CodeCleanup/AbstractCodeCleanUpFixer.cs

```diff
@@ -233,87 +206,68 @@ async Task<Solution> ApplyFixAsync(IProgress<CodeAnalysisProgress> progress, Can
-
-        return solution;
-    }
-
-    private async Task<Project> FixProjectAsync(
-        Project project,
-        FixIdContainer enabledFixIds,
-        IProgress<CodeAnalysisProgress> progressTracker,
-        bool addProgressItemsForDocuments,
-        CancellationToken cancellationToken)
-    {
-        if (!CanCleanupProject(project))
-        {
-            return project;
-        }
-
-        if (addProgressItemsForDocuments)
-        {
-            progressTracker.AddItems(project.DocumentIds.Count);
-        }
-
-        var ideOptions = _globalOptions.GetCodeActionOptions(project.Services);
-
-        foreach (var documentId in project.DocumentIds)
-        {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            var document = project.GetRequiredDocument(documentId);
-            progressTracker.Report(CodeAnalysisProgress.Description(document.Name));
+        // Add an item for each document in all the projects we're processing.
+        progressTracker.AddItems(projects.Sum(static p => p.DocumentIds.Count));
 
-            // FixDocumentAsync reports progress within a document, but we only want to report progress at the
-            // project granularity.  So we pass CodeAnalysisProgress.None here so that inner progress updates don't
-            // affect us.
-            var fixedDocument = await FixDocumentAsync(document, enabledFixIds, CodeAnalysisProgress.None, ideOptions, cancellationToken).ConfigureAwait(false);
-            project = fixedDocument.Project;
-            progressTracker.ItemCompleted();
-        }
-
-        return project;
+        // Run in parallel across all projects.
+        return await ProducerConsumer<(DocumentId documentId, SyntaxNode newRoot)>.RunParallelAsync(
+            source: projects,
+            produceItems: static async (project, callback, args, cancellationToken) =>
+            {
+                Contract.ThrowIfFalse(project.SupportsCompilation);
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var ideOptions = args.globalOptions.GetCodeActionOptions(project.Services);
+
+                // And for each project, process all the documents in parallel.
+                await RoslynParallel.ForEachAsync(
+                    source: project.Documents,
+                    cancellationToken,
+                    async (document, cancellationToken) =>
+                    {
+                        using var _ = args.progressTracker.ItemCompletedScope();
+
+                        // FixDocumentAsync reports progress within a document, but we only want to report progress at
+                        // the document granularity.  So we pass CodeAnalysisProgress.None here so that inner progress
+                        // updates don't affect us.
+                        var fixedDocument = await FixDocumentAsync(document, args.enabledFixIds, CodeAnalysisProgress.None, ideOptions, cancellationToken).ConfigureAwait(false);
```

Comment:
note that this logic is different from before.  now we operate in parallel, always processing documents from the *original* solution (previously we would process a document, fork the solution, and repeat).  This was obviously insanely slow.